moebot_bot/util/rolerules/roleRule.go

```diff
@@ -0,0 +1,68 @@
+package rolerules
+
+import (
+	"log"
+
+	"github.com/bwmarrin/discordgo"
+	"github.com/camd67/moebot/moebot_bot/util/db"
+	"github.com/camd67/moebot/moebot_bot/util/db/types"
+)
+
+//RoleActionType Type of action being performed on the role
+type RoleActionType int
+
+//RoleAction Action being performed on the role
+type RoleAction struct {
+	Role            *types.Role
+	UserRank        *types.UserServerRank
+	Member          *discordgo.Member
+	Guild           *discordgo.Guild
+	Channel         *discordgo.Channel
+	OriginalMessage *discordgo.Message
+	Action          RoleActionType
+}
+
+const (
+	//RoleAdd Adding the role
+	RoleAdd RoleActionType = 1
+	//RoleRemove Removing the role
+	RoleRemove RoleActionType = 2
+)
+
+//RoleRule Rule defining if and how a role should be applied
+type RoleRule interface {
+	Check(session *discordgo.Session, action *RoleAction) (success bool, message string)
+	Apply(session *discordgo.Session, action *RoleAction) (success bool, message string)
+}
+
+func GetRulesForRole(server *types.Server, role *types.Role, comPrefix string) ([]RoleRule, error) {
+	var result []RoleRule
+	if server.VeteranRole.String == role.RoleUid && server.VeteranRank.Valid {
+		result = append(result, &Points{PointsTreshold: int(server.VeteranRank.Int64)})
+	}
+	if role.ConfirmationMessage.Valid {
+		result = append(result, &Confirmation{ComPrefix: comPrefix})
+	}
+	for _, gID := range role.Groups {
+		group, err := db.RoleGroupQueryId(gID)
+		if err != nil {
+			log.Println("Error while retrieving role group during rules initialization", err)
+			return nil, err
+		}
+		relatedRoles, err := db.RoleQueryGroup(gID)
+		if err != nil {
+			log.Println("Error while retrieving related group roles during rules initialization", err)
+			return nil, err
+		}
+		if group.Type == types.GroupTypeExclusive || group.Type == types.GroupTypeExclusiveNoRemove {
+			result = append(result, &Exclusive{ExclusiveRoles: relatedRoles})
+		}
+		if group.Type == types.GroupTypeExclusiveNoRemove {
```

Comment:
The code was trying to reflect that we can have multiple roles for every group, and that some groups share rules. I've changed it to a switch, but let's keep an eye on that piece of code, it could become too complex or with lots of duplicated initializations in the future.

Solved in https://github.com/camd67/moebot/pull/67/commits/69b269d7a0aec8141240e393496f8e74aa72b617